File: mobile/ios/LiveActivity/RecordingActivityModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(RecordingActivityModule, NSObject)

RCT_EXTERN_METHOD(start:(NSString *)sessionId
                  title:(NSString *)title
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(update:(BOOL)isRecording
                  elapsedSeconds:(nonnull NSNumber *)elapsedSeconds
                  title:(NSString *)title
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stop:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: mobile/ios/LiveActivity/RecordingActivityModule.swift
import Foundation
import React
import ActivityKit

@objc(RecordingActivityModule)
class RecordingActivityModule: NSObject {

    @objc static func requiresMainQueueSetup() -> Bool { true }

    @objc(start:title:resolver:rejecter:)
    func start(sessionId: String,
               title: String,
               resolver: @escaping RCTPromiseResolveBlock,
               rejecter: @escaping RCTPromiseRejectBlock) {
        guard #available(iOS 16.2, *) else {
            rejecter("UNSUPPORTED", "iOS < 16.2", nil)
            return
        }

        do {
            try RecordingLiveActivityManager.shared.start(
                sessionId: sessionId,
                title: title
            )
            resolver(nil)
        } catch {
            rejecter("START_ERROR", error.localizedDescription, error)
        }
    }

    @objc(update:elapsedSeconds:title:resolver:rejecter:)
    func update(isRecording: Bool,
                elapsedSeconds: NSNumber,
                title: String,
                resolver: @escaping RCTPromiseResolveBlock,
                rejecter: @escaping RCTPromiseRejectBlock) {
        guard #available(iOS 16.2, *) else {
            rejecter("UNSUPPORTED", "iOS < 16.2", nil)
            return
        }

        RecordingLiveActivityManager.shared.update(
            isRecording: isRecording,
            elapsedSeconds: elapsedSeconds.intValue,
            title: title
        )
        resolver(nil)
    }

    @objc(stop:rejecter:)
    func stop(resolver: @escaping RCTPromiseResolveBlock,
              rejecter: @escaping RCTPromiseRejectBlock) {
        guard #available(iOS 16.2, *) else {
            rejecter("UNSUPPORTED", "iOS < 16.2", nil)
            return
        }

        RecordingLiveActivityManager.shared.end()
        resolver(nil)
    }
}
